tests/components/mqtt/test_init.py

```diff
@@ -64,619 +58,642 @@ def mock_mqtt():
+    async_fire_mqtt_message(hass, topic, payload)
+    await hass.async_block_till_done()
+    assert len(calls) == 1
+    assert calls[0][0].topic == topic
+    assert calls[0][0].payload == payload
 
-        def side_effect(*args):
-            async_fire_mqtt_message(self.hass, "test/state", "online")
-            return 0, 0
 
-        self.hass.data["mqtt"]._mqttc.subscribe.side_effect = side_effect
+async def test_retained_message_on_subscribe_received(
+    hass, mqtt_client_mock, mqtt_mock
+):
+    """Test every subscriber receives retained message on subscribe."""
 
-        # Fake that the client is connected
-        self.hass.data["mqtt"].connected = True
+    def side_effect(*args):
+        async_fire_mqtt_message(hass, "test/state", "online")
+        return 0, 0
 
-        calls_a = MagicMock()
-        mqtt.subscribe(self.hass, "test/state", calls_a)
-        self.hass.block_till_done()
-        assert calls_a.called
+    mqtt_client_mock.subscribe.side_effect = side_effect
 
-        calls_b = MagicMock()
-        mqtt.subscribe(self.hass, "test/state", calls_b)
-        self.hass.block_till_done()
-        assert calls_b.called
+    # Fake that the client is connected
+    mqtt_mock.connected = True
 
-    def test_not_calling_unsubscribe_with_active_subscribers(self):
-        """Test not calling unsubscribe() when other subscribers are active."""
-        # Fake that the client is connected
-        self.hass.data["mqtt"].connected = True
+    calls_a = MagicMock()
+    await mqtt.async_subscribe(hass, "test/state", calls_a)
+    await hass.async_block_till_done()
+    assert calls_a.called
 
-        unsub = mqtt.subscribe(self.hass, "test/state", None)
-        mqtt.subscribe(self.hass, "test/state", None)
-        self.hass.block_till_done()
-        assert self.hass.data["mqtt"]._mqttc.subscribe.called
+    calls_b = MagicMock()
+    await mqtt.async_subscribe(hass, "test/state", calls_b)
+    await hass.async_block_till_done()
+    assert calls_b.called
 
-        unsub()
-        self.hass.block_till_done()
-        assert not self.hass.data["mqtt"]._mqttc.unsubscribe.called
 
-    def test_restore_subscriptions_on_reconnect(self):
-        """Test subscriptions are restored on reconnect."""
-        # Fake that the client is connected
-        self.hass.data["mqtt"].connected = True
+async def test_not_calling_unsubscribe_with_active_subscribers(
+    hass, mqtt_client_mock, mqtt_mock
+):
+    """Test not calling unsubscribe() when other subscribers are active."""
+    # Fake that the client is connected
+    mqtt_mock.connected = True
 
-        mqtt.subscribe(self.hass, "test/state", None)
-        self.hass.block_till_done()
-        assert self.hass.data["mqtt"]._mqttc.subscribe.call_count == 1
+    unsub = await mqtt.async_subscribe(hass, "test/state", None)
+    await mqtt.async_subscribe(hass, "test/state", None)
+    await hass.async_block_till_done()
+    assert mqtt_client_mock.subscribe.called
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5737505)